graphsync.go

```diff
@@ -98,7 +98,35 @@ type ResponseProgress struct {
 	}
 }
 
+// OnRequestReceivedHook processes extension data in a request
+// responseData - the value that should be sent for the extension in the reply
+// err - error - if not nil, halt request and return RequestRejected with the responseData
+type OnRequestReceivedHook func(p peer.ID, root ipld.Link, selector ipld.Node, requestData []byte) (responseData ExtensionData, err error)
+
+// OnResponseReceivedHook processes extension data in a response
+// When it returns an error processing is halted and the original request is cancelled
+type OnResponseReceivedHook func(p peer.ID, responseData []byte) error
+
+// ExtensionConfig defines behavior for user supplied extension
+type ExtensionConfig struct {
+	Name ExtensionName
+	// PerformsValidation specifies if the request received hook can be considered "a validator"
+	// if true, and the hook does not error, then one can assume the request is valid and this
+	// should override default validation schemes
+	PerformsValidation bool
```

Comment:
So the use case is this: by default, Graphsync will limit recursive selectors to a maximum depth (100 for now). In order for a single graphsync request to transfer the whole graph in the case of a large piece, the Data transfer layer uses an extension which checks the request against storage deal proposal -- which verifies that both parties have already agreed to this transfer, and we are in a context we can ignore the selector recursion limit. We could disable the limit globally, BUT we can only run 1 Graphsync instance on a graphsync node. We definitely DON'T want to disable the limit for something like BlockSync. To me this is somewhat like sending a cookie -- there are things you can do in one context where you can't in another.

There are interesting security implications -- if we don't verify peer identity that could be a problem. It's possible we need to re-sign the data that goes into the data transfer extension (instead of just encoding the proposal CID). Perhaps this is a question for a security oriented person. But definitiely otherwise there's no way we can do unlimited recursion selectors.

Maybe PerformsAuthentication + A Described Permission level is better?